<a href="https://colab.research.google.com/github/mihnguyen/udemy-customer-analytics/blob/main/purchase-prediction/notebook_Purchase_Prediction_Pt2_Predictive_Modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This notebook is part 2 and a continuation of the [[Udemy CAiP] Purchase Prediction Pt1 - Descriptive Analysis notebook](https://colab.research.google.com/drive/1Acmi4BOuO1V7LCxbnMCVzu46vFSpftA6#scrollTo=97I0svM77BjK&uniqifier=1).
While part 1 is descriptive analysis, part 2 will concern predictive modeling.

###Import Libraries

In [1]:
import numpy as np
import pandas as pd

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans

import pickle

from sklearn.linear_model import LogisticRegression

###Import Purchase Data

In [2]:
from google.colab import files
uploaded = files.upload()

Saving purchase data.csv to purchase data.csv


In [3]:
import io
df_purchase = pd.read_csv(io.BytesIO(uploaded['purchase data.csv']))

###Import Pickle Objects

In [5]:
# Import pickle objects
from google.colab import drive
drive.mount('/content/drive')

import pickle
drive.mount('/content/drive')
DATA_PATH = "/content/drive/My Drive/Colab Notebooks"

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
infile = open(DATA_PATH+'/pca.pickle','rb')
pca = pickle.load(infile)

pca

PCA(copy=True, iterated_power='auto', n_components=3, random_state=None,
    svd_solver='auto', tol=0.0, whiten=False)

In [7]:
infile = open(DATA_PATH+'/scaler.pickle','rb')
scaler = pickle.load(infile)

scaler

StandardScaler(copy=True, with_mean=True, with_std=True)

In [8]:
infile = open(DATA_PATH+'/kmeans_pca.pickle','rb')
kmeans_pca = pickle.load(infile)

kmeans_pca

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=4, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=42, tol=0.0001, verbose=0)

In [9]:
# Standardize data
features = df_purchase[['Sex', 'Marital status', 'Age', 'Education', 'Income', 'Occupation', 'Settlement size']]
df_purchase_segm_std = scaler.transform(features)

# Apply PCA
df_purchase_segm_pca = pca.transform(df_purchase_segm_std)

# Segment data
purchase_segm_kmeans_pca = kmeans_pca.predict(df_purchase_segm_pca)

# Create copy of dataframe
df_purchase_predictors = df_purchase.copy()

# Add segment labels
df_purchase_predictors['Segment'] = purchase_segm_kmeans_pca
segment_dummies = pd.get_dummies(purchase_segm_kmeans_pca, prefix = 'Segment', prefix_sep = '_')
df_purchase_predictors = pd.concat([df_purchase_predictors, segment_dummies], axis = 1)

df_pa = df_purchase_predictors

###Modeling Purchase Probability

Now we will calculate the probability of a customer purchasing at each shop visit. Method used is Logistic Regression.

First, define the dependent and independent variables.

Dependent variable is whether a purchase happens or not (values are either 0 or 1).

As for independent variable, common sense is that price has the most correlation to purchase. So price is the first independent variable. Here let's use the mean price (of all 5 brands) as the independent variable:

In [10]:
Y = df_pa['Incidence']

In [11]:
X = pd.DataFrame()
X['Mean Price'] = (df_pa['Price_1'] +
                   df_pa['Price_2'] +
                   df_pa['Price_3'] +
                   df_pa['Price_4'] +
                   df_pa['Price_5'] ) / 5

Now fit model to predic purchase probability:

In [12]:
model_purchase = LogisticRegression(solver = 'sag') # 'sag' was chosen as the solver as it's optimal for simple problem with large dataset like this one
model_purchase.fit(X, Y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='sag', tol=0.0001, verbose=0,
                   warm_start=False)

Now check the exact relationship between (Mean) Price (X) and Purchase Probability (Y) using coefficient for the independent variable (coef_ attribute), which is the mean price:

In [13]:
model_purchase.coef_

array([[-2.34802882]])

The coefficient for mean price = -2.35. This indicates that an increase in price will result in a decrease in purchase probability (obviously). However, this coefficient only shows the direction of the relationship (which in this case is reverse, or negative).

Next, let's quantify the magnitude of the negative relationship between mean price and purchase probability.

###Modeling Price Elasticity of Purchase Probability

The purpose of this modeling is to check the effect of various mean prices on the purchase probability. Various mean prices here can be done by generating a range of mean price values. How to determine that range?

First, by checking the 5 prices (Price_1 to Price_5) to see their distributions. This checking can be done through .describe() function:

In [14]:
df_pa[['Price_1', 'Price_2', 'Price_3', 'Price_4', 'Price_5']].describe()

Price_1       Price_2       Price_3       Price_4       Price_5
count  58693.000000  58693.000000  58693.000000  58693.000000  58693.000000
mean       1.392074      1.780999      2.006789      2.159945      2.654798
std        0.091139      0.170868      0.046867      0.089825      0.098272
min        1.100000      1.260000      1.870000      1.760000      2.110000
25%        1.340000      1.580000      1.970000      2.120000      2.630000
50%        1.390000      1.880000      2.010000      2.170000      2.670000
75%        1.470000      1.890000      2.060000      2.240000      2.700000
max        1.590000      1.900000      2.140000      2.260000      2.800000

The minimum price is 1.10 (min Price_1), while the maximum price is 2.80 (max Price_5).

Then, the range of mean price we could pick is, say 0.50 to 3.50, for example. Now let's generate a range of numbers between 0.5 and 3.5, with steps = 0.01 to denote an increase in 1 cent (an increase in 1% of a dollar):

In [15]:
price_range = np.arange(0.5, 3.5, 0.01)
price_range

array([0.5 , 0.51, 0.52, 0.53, 0.54, 0.55, 0.56, 0.57, 0.58, 0.59, 0.6 ,
       0.61, 0.62, 0.63, 0.64, 0.65, 0.66, 0.67, 0.68, 0.69, 0.7 , 0.71,
       0.72, 0.73, 0.74, 0.75, 0.76, 0.77, 0.78, 0.79, 0.8 , 0.81, 0.82,
       0.83, 0.84, 0.85, 0.86, 0.87, 0.88, 0.89, 0.9 , 0.91, 0.92, 0.93,
       0.94, 0.95, 0.96, 0.97, 0.98, 0.99, 1.  , 1.01, 1.02, 1.03, 1.04,
       1.05, 1.06, 1.07, 1.08, 1.09, 1.1 , 1.11, 1.12, 1.13, 1.14, 1.15,
       1.16, 1.17, 1.18, 1.19, 1.2 , 1.21, 1.22, 1.23, 1.24, 1.25, 1.26,
       1.27, 1.28, 1.29, 1.3 , 1.31, 1.32, 1.33, 1.34, 1.35, 1.36, 1.37,
       1.38, 1.39, 1.4 , 1.41, 1.42, 1.43, 1.44, 1.45, 1.46, 1.47, 1.48,
       1.49, 1.5 , 1.51, 1.52, 1.53, 1.54, 1.55, 1.56, 1.57, 1.58, 1.59,
       1.6 , 1.61, 1.62, 1.63, 1.64, 1.65, 1.66, 1.67, 1.68, 1.69, 1.7 ,
       1.71, 1.72, 1.73, 1.74, 1.75, 1.76, 1.77, 1.78, 1.79, 1.8 , 1.81,
       1.82, 1.83, 1.84, 1.85, 1.86, 1.87, 1.88, 1.89, 1.9 , 1.91, 1.92,
       1.93, 1.94, 1.95, 1.96, 1.97, 1.98, 1.99, 2.

That's 300 generated values for mean price.

Now store this range in dataframe called df_price_range:

In [16]:
df_price_range = pd.DataFrame(price_range)

Predict the purchase probability for each mean price (each of the generated prices above):

In [17]:
Y_pr = model_purchase.predict_proba(df_price_range)
Y_pr

array([[0.08204028, 0.91795972],
       [0.08382601, 0.91617399],
       [0.08564699, 0.91435301],
       [0.08750375, 0.91249625],
       [0.08939682, 0.91060318],
       [0.09132675, 0.90867325],
       [0.09329408, 0.90670592],
       [0.09529934, 0.90470066],
       [0.09734308, 0.90265692],
       [0.09942583, 0.90057417],
       [0.10154812, 0.89845188],
       [0.1037105 , 0.8962895 ],
       [0.1059135 , 0.8940865 ],
       [0.10815765, 0.89184235],
       [0.11044348, 0.88955652],
       [0.1127715 , 0.8872285 ],
       [0.11514225, 0.88485775],
       [0.11755623, 0.88244377],
       [0.12001395, 0.87998605],
       [0.12251593, 0.87748407],
       [0.12506265, 0.87493735],
       [0.12765461, 0.87234539],
       [0.13029229, 0.86970771],
       [0.13297616, 0.86702384],
       [0.13570669, 0.86429331],
       [0.13848434, 0.86151566],
       [0.14130954, 0.85869046],
       [0.14418273, 0.85581727],
       [0.14710433, 0.85289567],
       [0.15007475, 0.84992525],
       [0.

That's an array with 300 rows (corresponding to 300 generated values for mean price), and 2 columns: the first column indicates the probability of no purchase, the second column the probability of purchase. Obviously, the lower the price, the higher the purchase probability.

Let's store it in a variable called purchase_pr:

In [18]:
purchase_pr = Y_pr[:][:, 1]

Now, in order to answer how purchase probability changes with a given change in price, we'll calculate the price elasticities. Let's recall the definition of price elasticity:

<img src="https://drive.google.com/uc?id=11cndOhtTFcAknIIhpE30UaG-2qmhpiHH" height="100">

The simplified formula for the formula above is:
E = beta * price * (1 - Pr(purchase))

Following that simplified formula, let's create the variable price_elasticity that calculates the price elasticity at each price point: 
- price_elasticity is an array with shape of 1 x number of features. In this case, the number of features = 1. Therefore, to access the price coefficient, the index is [0].
- The price is contained in price_range variable.
- Pr(purchase) is purchase_pr.

In [19]:
price_elasticity = model_purchase.coef_[:, 0] * price_range * (1 - purchase_pr)

Now, let's create the dataframe df_price_elasticities to store all price elasticities:

In [20]:
df_price_elasticities = pd.DataFrame(price_range)

In [22]:
df_price_elasticities = df_price_elasticities.rename(columns = {0: "Price Point"})
df_price_elasticities['Mean Price Elasticity'] = price_elasticity

pd.options.display.max_rows = None # to see all rows
df_price_elasticities

Price Point  Mean Price Elasticity
0           0.50              -0.096316
1           0.51              -0.100381
2           0.52              -0.104573
3           0.53              -0.108894
4           0.54              -0.113349
5           0.55              -0.117941
6           0.56              -0.122672
7           0.57              -0.127546
8           0.58              -0.132567
9           0.59              -0.137738
10          0.60              -0.143063
11          0.61              -0.148544
12          0.62              -0.154187
13          0.63              -0.159993
14          0.64              -0.165968
15          0.65              -0.172114
16          0.66              -0.178436
17          0.67              -0.184937
18          0.68              -0.191621
19          0.69              -0.198493
20          0.70              -0.205555
21          0.71              -0.212813
22          0.72              -0.220270
23          0.73              -0.227929
24          0.74              -0.235796
25          0.75              -0.243874
26          0.76              -0.252167
27          0.77              -0.260680
28          0.78              -0.269416
29          0.79              -0.278380
30          0.80              -0.287576
31          0.81              -0.297008
32          0.82              -0.306681
33          0.83              -0.316597
34          0.84              -0.326763
35          0.85              -0.337182
36          0.86              -0.347857
37          0.87              -0.358794
38          0.88              -0.369997
39          0.89              -0.381469
40          0.90              -0.393214
41          0.91              -0.405238
42          0.92              -0.417543
43          0.93              -0.430134
44          0.94              -0.443015
45          0.95              -0.456190
46          0.96              -0.469662
47          0.97              -0.483436
48          0.98              -0.497515
49          0.99              -0.511902
50          1.00              -0.526603
51          1.01              -0.541619
52          1.02              -0.556955
53          1.03              -0.572613
54          1.04              -0.588598
55          1.05              -0.604912
56          1.06              -0.621558
57          1.07              -0.638540
58          1.08              -0.655860
59          1.09              -0.673520
60          1.10              -0.691524
61          1.11              -0.709874
62          1.12              -0.728572
63          1.13              -0.747620
64          1.14              -0.767021
65          1.15              -0.786775
66          1.16              -0.806886
67          1.17              -0.827354
68          1.18              -0.848181
69          1.19              -0.869368
70          1.20              -0.890916
71          1.21              -0.912826
72          1.22              -0.935099
73          1.23              -0.957734
74          1.24              -0.980733
75          1.25              -1.004094
76          1.26              -1.027820
77          1.27              -1.051907
78          1.28              -1.076358
79          1.29              -1.101169
80          1.30              -1.126342
81          1.31              -1.151873
82          1.32              -1.177763
83          1.33              -1.204009
84          1.34              -1.230610
85          1.35              -1.257563
86          1.36              -1.284866
87          1.37              -1.312517
88          1.38              -1.340514
89          1.39              -1.368852
90          1.40              -1.397530
91          1.41              -1.426543
92          1.42              -1.455889
93          1.43              -1.485563
94          1.44              -1.515562
95          1.45              -1.545881
96          1.46              -1.576516
97          1.47              -1.607463
98          1.48              -1.638717
99   